### DISCUSSION TITLE
osu! touch input improvements

### DISCUSSION THREAD

**Discussion by Walavouchey** [ppy/osu#22886] [Ideas]
alternative to https://github.com/ppy/osu/pull/22873, addresses https://github.com/ppy/osu/issues/22556.

branch: https://github.com/walavouchey/osu/tree/osu-touch-input

reply from @peppy: https://github.com/ppy/osu/pull/22873#discussion_r1139978040

> It severely deprives the user feedback loop for non-offscreen-tapping gameplay (in the case of tracking a slider; in the case of missing an object; etc.)

i have no idea what this means. have you tested gameplay with the branch, and confirmed that it behaves way off expectations? the cases you mention behave exactly the same in my branch as in yours.

demo (first tapping off screen, then tapping directly on hit objects):

https://user-images.githubusercontent.com/36758269/225894017-a1f5c918-6b7c-44aa-8523-427b1e8b6b96.mp4

notice how the game immediately picks up the correct finger to track at the start of the map, since that one is being held.

for a demo on peppy's pr branch, see https://github.com/ppy/osu/pull/22873#discussion_r1139046496
for a demo on latest release, see https://github.com/ppy/osu/issues/22556#issuecomment-1431492733

another demo (always keeping at least one finger held):

https://user-images.githubusercontent.com/36758269/225899888-3d28be43-4c31-4e16-88ab-ce91e80aa4bd.mp4

in any case i believe this is more of a correct direction, both tapping styles considered. not doing things this way makes off-screen touch input feel awkward.

**Comment by peppy** [comment]
Try tapping just off an object (missing positionally) while having another finger on screen.

I'd hope these kinds of cases are pretty obvious no?

You can run through the test scene for assumptions.

**Reply by Walavouchey** [reply]
if you do that, the click should happen at the point of the held finger, no?

you either miss the object close to the tapping finger, or you intended to hit a stack under the held finger by tapping with your offhand, akin to how osu!droid players play the game. it's more intuitive and doesn't require weird workarounds.

**Reply by Walavouchey** [reply]
even if you keep one finger held for some reason and continue to tap normally with one hand, it'll ignore the held finger for subsequent taps.

in case this is what you meant, it already works this way on my branch, as my change doesn't affect that part:

https://user-images.githubusercontent.com/36758269/225967686-e9c134d3-39dc-414d-a6af-6c9570007df7.mp4

**Reply by peppy** [reply]
- Touch circle (successfully)
- Without releasing, miss next circle

**Reply by Walavouchey** [reply]
> * Touch circle (successfully)
> * Without releasing, miss next circle

after testing, both branches behave exactly the same (second tap activates click action at the location of the first finger, as expected).

i still have no idea what you're trying to say and you seem to be ignoring the core issue. i'll revisit this in a few weeks when more people have played on the next release.

**Reply by peppy** [reply]
There are comprehensive tests you can run to see how your changes change the expectations I've set in place:

![JetBrains Rider-EAP 2023-03-21 at 17 42 51](https://user-images.githubusercontent.com/191335/226696421-ed1fa4a1-737e-4ab3-8736-cdc4b07ee85c.png)

Your change is missing important implementation details. You are likely missing them due to not looking at the key counter and not testing the specific scenario I proposed (your assertion in your reply doesn't match my expectations). Tests are good for helping to understand this.

**Reply by Walavouchey** [reply]
I'm not expecting my changes to pass the tests as i haven't checked those corner cases. I checked that gameplay works intuitively. The point wasn't to get this merged as-is, but to point out that the current implementation is flawed and to suggest a more correct direction for the basic touch input behaviour.

> in any case i believe this is more of a correct direction, both tapping styles considered. not doing things this way makes off-screen touch input feel awkward.

So far you've been implying that you'd be against this direction:

> It's not the direction I want to take sorry. It severely deprives the user feedback loop for non-offscreen-tapping gameplay (in the case of tracking a slider; in the case of missing an object; etc.)

I can add changes to match relevant test expectations and update any tests as required, but with the immediate resistance to the very idea along with seemingly irrelevant counterpoints, I'm under the impression that I need to convince you that what we have right now is very jank.

Maybe I should rephrase myself: would you agree that the behaviour pointed out in https://github.com/ppy/osu/pull/22873#discussion_r1139046496 is jank? (as caused by the lines i removed in my branch)

Again, I'll revisit this in a few weeks (I also don't have the bandwidth to make my branch into a pr-able state rn.)

**Reply by Walavouchey** [reply]
> not testing the specific scenario I proposed (your assertion in your reply doesn't match my expectations)

I'm not sure what you mean here. I don't know what else you'd expect based on the code.

> > * Touch circle (successfully)
> > * Without releasing, miss next circle
> 
> after testing, both branches behave exactly the same (second tap activates click action at the location of the first finger, as expected).

If you touch a circle and keep holding, it'll immediately be [tracked as positional input](https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Osu/UI/OsuTouchInputMapper.cs#L93-L98). When you then tap outside of any circle, it'll cause a click under the already tracked finger (and not set a new position-tracked finger), all according to the expectations I laid out earlier https://github.com/ppy/osu/discussions/22886#discussioncomment-5346056

That's beside the point though, as the change in my branch only affects multi-touch when not tapping directly on any hit objects. This scenario you described should behave the same as the case when not first tapping on a hit object, which is what my proposed change is addressing.
